src/client/cli/formatter/json_formatter.cpp

```diff
@@ -63,90 +63,117 @@ std::string mp::JsonFormatter::format(const InfoReply& reply) const
+            instance_info.insert("memory", memory);
+
+            QJsonArray ipv4_addrs;
+            for (const auto& ip : instance_details.ipv4())
+                ipv4_addrs.append(QString::fromStdString(ip));
+            instance_info.insert("ipv4", ipv4_addrs);
+
+            QJsonObject mounts;
+            for (const auto& mount : info.mount_info().mount_paths())
             {
-                auto gid_map_pair = mount_maps.gid_mappings(i);
-                auto host_gid = gid_map_pair.host_id();
-                auto instance_gid = gid_map_pair.instance_id();
-
-                mount_gids.append(
-                    QString("%1:%2")
-                        .arg(QString::number(host_gid))
-                        .arg((instance_gid == mp::default_id) ? "default" : QString::number(instance_gid)));
+                QJsonObject entry;
+                QJsonArray mount_uids;
+                QJsonArray mount_gids;
+
+                auto mount_maps = mount.mount_maps();
+
+                for (auto i = 0; i < mount_maps.uid_mappings_size(); ++i)
+                {
+                    auto uid_map_pair = mount_maps.uid_mappings(i);
+                    auto host_uid = uid_map_pair.host_id();
+                    auto instance_uid = uid_map_pair.instance_id();
+
+                    mount_uids.append(
+                        QString("%1:%2")
+                            .arg(QString::number(host_uid))
+                            .arg((instance_uid == mp::default_id) ? "default" : QString::number(instance_uid)));
+                }
+                for (auto i = 0; i < mount_maps.gid_mappings_size(); ++i)
+                {
+                    auto gid_map_pair = mount_maps.gid_mappings(i);
+                    auto host_gid = gid_map_pair.host_id();
+                    auto instance_gid = gid_map_pair.instance_id();
+
+                    mount_gids.append(
+                        QString("%1:%2")
+                            .arg(QString::number(host_gid))
+                            .arg((instance_gid == mp::default_id) ? "default" : QString::number(instance_gid)));
+                }
+                entry.insert("uid_mappings", mount_uids);
+                entry.insert("gid_mappings", mount_gids);
+                entry.insert("source_path", QString::fromStdString(mount.source_path()));
+
+                mounts.insert(QString::fromStdString(mount.target_path()), entry);
             }
-            entry.insert("uid_mappings", mount_uids);
-            entry.insert("gid_mappings", mount_gids);
-            entry.insert("source_path", QString::fromStdString(mount.source_path()));
+            instance_info.insert("mounts", mounts);
+
+            info_obj.insert(QString::fromStdString(info.name()), instance_info);
+        }
+    }
+    else if (reply.info_contents_case() == mp::InfoReply::kSnapshotOverview)
+    {
+        std::unordered_map<std::string, QJsonObject> instance_map;
```

Comment:
The reason I had a separate map is because Qt JSON objects don't allow you insert into nested objects using a reference (like an iterator). And because I don't want to rely on the snapshot info being grouped together by instance, I had first grouped the info together in a map before inserting it all into the main `info_json` object.

I pushed an alternate approach, but it still has the downside of creating a new QJsonObject every time snapshot info has to be inserted for an instance that has already been seen in the loop.

Reply:
OK, got it. I think we can keep the latter approach.